quixstreams/sinks/community/postgresql.py

```diff
@@ -125,33 +156,50 @@ def setup(self):
 
     def write(self, batch: SinkBatch):
         tables = {}
+        primary_keys = []
+
         for item in batch:
             table = tables.setdefault(
-                self._table_name(item), {"rows": [], "cols_types": {}}
+                self._table_name(item), {"rows": [], "cols_types": {}, "pks_types": {}}
             )
+            if self._primary_keys:
+                primary_keys = self._primary_keys(item) or []
+                if isinstance(primary_keys, str):
+                    primary_keys = [primary_keys]
+
             row = {}
             if item.key is not None:
                 key_type = type(item.key)
-                table["cols_types"].setdefault(_KEY_COLUMN_NAME, key_type)
-                row[_KEY_COLUMN_NAME] = item.key
-
+                table["cols_types"].setdefault(KEY_COLUMN_NAME, key_type)
+                row[KEY_COLUMN_NAME] = item.key
             for key, value in item.value.items():
-                if value is not None:
+                if key in primary_keys:
+                    table["pks_types"].setdefault(key, type(key))
+                    row[key] = value
+                elif value is not None:
                     table["cols_types"].setdefault(key, type(value))
                     row[key] = value
 
-            row[_TIMESTAMP_COLUMN_NAME] = datetime.fromtimestamp(item.timestamp / 1000)
+            row[TIMESTAMP_COLUMN_NAME] = datetime.fromtimestamp(item.timestamp / 1000)
             table["rows"].append(row)
 
         try:
             with self._client:
                 for name, values in tables.items():
                     if self._schema_auto_update:
                         self._create_table(name)
+                        if self._primary_keys:
+                            self._add_new_primary_keys(name, values["pks_types"])
                         self._add_new_columns(name, values["cols_types"])
-                    self._insert_rows(name, values["rows"])
+                    self._insert_rows(name, values["rows"], values["pks_types"])
         except psycopg2.Error as e:
             self._client.rollback()
+            if "duplicate key value violates unique constraint" in str(e):
+                raise PostgreSQLSinkException(
+                    f"Failed to write batch: attempted insert of new row using an "
+                    f"existing primary key. Try `upsert_on_primary_key=True` for "
+                    f"upsert functionality. PostgreSQL error info: {str(e)}"
+                ) from e
             raise PostgreSQLSinkException(f"Failed to write batch: {str(e)}") from e
         table_counts = {table: len(values["rows"]) for table, values in tables.items()}
```

Comment:
Maybe there is a way to avoid iterating over `tables.items()` twice? You can initialize the `table_counts`  dict before the previous for loop and update it on the go.